File: backend/src/main/java/com/ecommerce/service/OrderService.java
package com.ecommerce.service;

import com.ecommerce.dto.*;
import com.ecommerce.entity.*;
import com.ecommerce.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class OrderService {
    private final OrderRepository orderRepository;
    private final CartRepository cartRepository;
    private final CartItemRepository cartItemRepository;
    private final CouponRepository couponRepository;
    private final UserRepository userRepository;
    private final ProductRepository productRepository;
    
    @Transactional
    public OrderDTO createOrder(Long userId, CreateOrderRequest request) {
        Cart cart = cartRepository.findByUserId(userId)
                .orElseThrow(() -> new RuntimeException("Cart not found"));
        
        if (cart.getItems().isEmpty()) {
            throw new RuntimeException("Cart is empty");
        }
        
        User user = userRepository.findById(userId).orElseThrow();
        
        BigDecimal subtotal = cart.getItems().stream()
                .map(item -> item.getProduct().getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        
        BigDecimal discount = BigDecimal.ZERO;
        Coupon coupon = null;
        if (request.getCouponCode() != null && !request.getCouponCode().isEmpty()) {
            coupon = couponRepository.findByCodeAndIsActiveTrue(request.getCouponCode())
                    .orElseThrow(() -> new RuntimeException("Invalid coupon"));
            if (!coupon.isValid()) {
                throw new RuntimeException("Coupon expired or limit reached");
            }
            discount = coupon.calculateDiscount(subtotal);
            coupon.setUsedCount(coupon.getUsedCount() + 1);
            couponRepository.save(coupon);
        }
        
        BigDecimal shippingFee = subtotal.compareTo(BigDecimal.valueOf(500000)) >= 0 
                ? BigDecimal.ZERO : BigDecimal.valueOf(30000);
        BigDecimal total = subtotal.subtract(discount).add(shippingFee);
        
        Order order = Order.builder()
                .user(user)
                .orderNumber(generateOrderNumber())
                .subtotal(subtotal)
                .discountAmount(discount)
                .shippingFee(shippingFee)
                .totalAmount(total)
                .coupon(coupon)
                .paymentMethod(Order.PaymentMethod.valueOf(request.getPaymentMethod()))
                .shippingName(request.getShippingName())
                .shippingPhone(request.getShippingPhone())
                .shippingAddress(request.getShippingAddress())
                .note(request.getNote())
                .build();
        
        for (CartItem cartItem : cart.getItems()) {
            Product product = cartItem.getProduct();
            OrderItem orderItem = OrderItem.builder()
                    .order(order)
                    .product(product)
                    .productName(product.getName())
                    .productImage(product.getImageUrl())
                    .price(product.getPrice())
                    .quantity(cartItem.getQuantity())
                    .subtotal(product.getPrice().multiply(BigDecimal.valueOf(cartItem.getQuantity())))
                    .build();
            order.getItems().add(orderItem);
            
            product.setStockQuantity(product.getStockQuantity() - cartItem.getQuantity());
            product.setSoldQuantity(product.getSoldQuantity() + cartItem.getQuantity());
            productRepository.save(product);
        }
        
        orderRepository.save(order);
        cartItemRepository.deleteByCartId(cart.getId());
        
        return toDTO(order);
    }
    
    public Page<OrderDTO> getUserOrders(Long userId, int page, int size) {
        return orderRepository.findByUserIdOrderByCreatedAtDesc(userId, PageRequest.of(page, size))
                .map(this::toDTO);
    }
    
    public OrderDTO getOrderById(Long orderId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found"));
        return toDTO(order);
    }
    
    @Transactional
    public OrderDTO cancelOrder(Long orderId, Long userId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found"));
        
        if (!order.getUser().getId().equals(userId)) {
            throw new RuntimeException("Unauthorized");
        }
        if (order.getStatus() != Order.OrderStatus.PENDING) {
            throw new RuntimeException("Cannot cancel this order");
        }
        
        order.setStatus(Order.OrderStatus.CANCELLED);
        for (OrderItem item : order.getItems()) {
            Product product = item.getProduct();
            product.setStockQuantity(product.getStockQuantity() + item.getQuantity());
            product.setSoldQuantity(product.getSoldQuantity() - item.getQuantity());
            productRepository.save(product);
        }
        
        return toDTO(orderRepository.save(order));
    }
    
    @Transactional
    public OrderDTO updateOrderStatus(Long orderId, String status) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new RuntimeException("Order not found"));
        
        try {
            Order.OrderStatus newStatus = Order.OrderStatus.valueOf(status);
            order.setStatus(newStatus);
            
            // Nếu chuyển sang DELIVERED, cập nhật payment status
            if (newStatus == Order.OrderStatus.DELIVERED) {
                order.setPaymentStatus(Order.PaymentStatus.PAID);
            }
            
            return toDTO(orderRepository.save(order));
        } catch (IllegalArgumentException e) {
            throw new RuntimeException("Invalid order status: " + status);
        }
    }
    
    private String generateOrderNumber() {
        return "ORD" + LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyyMMddHHmmss")) 
                + UUID.randomUUID().toString().substring(0, 4).toUpperCase();
    }
    
    private OrderDTO toDTO(Order order) {
        var items = order.getItems().stream().map(item ->
            OrderItemDTO.builder()
                .id(item.getId())
                .productId(item.getProduct().getId())
                .productName(item.getProductName())
                .productImage(item.getProductImage())
                .price(item.getPrice())
                .quantity(item.getQuantity())
                .subtotal(item.getSubtotal())
                .build()
        ).collect(Collectors.toList());
        
        return OrderDTO.builder()
                .id(order.getId())
                .orderNumber(order.getOrderNumber())
                .status(order.getStatus().name())
                .subtotal(order.getSubtotal())
                .discountAmount(order.getDiscountAmount())
                .shippingFee(order.getShippingFee())
                .totalAmount(order.getTotalAmount())
                .paymentMethod(order.getPaymentMethod().name())
                .paymentStatus(order.getPaymentStatus().name())
                .shippingName(order.getShippingName())
                .shippingPhone(order.getShippingPhone())
                .shippingAddress(order.getShippingAddress())
                .note(order.getNote())
                .couponCode(order.getCoupon() != null ? order.getCoupon().getCode() : null)
                .items(items)
                .createdAt(order.getCreatedAt())
                .build();
    }
}

File: lib/services/order_service.dart
import '../config/api_config.dart';
import '../models/order.dart';
import 'api_service.dart';

class OrderService {
  // Tạo đơn hàng mới
  static Future<Map<String, dynamic>> createOrder({
    required List<Map<String, dynamic>> items,
    required String shippingName,
    required String shippingPhone,
    required String shippingAddress,
    required String paymentMethod,
    String? couponCode,
    double discount = 0,
    double shippingFee = 0,
    double total = 0,
  }) async {
    try {
      print('📦 [OrderService] Creating order...');
      print('💰 Total to send: $total');
      
      final response = await ApiService.post(
        ApiConfig.orders,
        {
          'items': items,
          'shippingName': shippingName,
          'shippingPhone': shippingPhone,
          'shippingAddress': shippingAddress,
          'paymentMethod': paymentMethod,
          'couponCode': couponCode,
          'discount': discount,
          'shippingFee': shippingFee,
          'total': total,
        },
        auth: true,
      );
      
      print('✅ [OrderService] Order created successfully');
      return response;
    } catch (e) {
      print('❌ [OrderService] Error creating order: $e');
      rethrow;
    }
  }

  // Lấy danh sách đơn hàng của user
  static Future<List<Order>> getMyOrders({int page = 0, int size = 20}) async {
    try {
      print('📦 [OrderService] Fetching my orders...');
      
      final response = await ApiService.get(
        '${ApiConfig.orders}?page=$page&size=$size',
        auth: true,
      );
      
      print('📊 [OrderService] Response: $response');
      
      if (response['success'] == true && response['data'] != null) {
        final data = response['data'];
        
        // Backend trả về Page object với content array
        if (data is Map && data['content'] != null) {
          final List<dynamic> content = data['content'];
          print('✅ [OrderService] Orders count: ${content.length}');
          return content.map((json) => Order.fromJson(json)).toList();
        }
        // Hoặc trả về trực tiếp array
        else if (data is List) {
          print('✅ [OrderService] Orders count: ${data.length}');
          return data.map((json) => Order.fromJson(json)).toList();
        }
      }
      print('⚠️ [OrderService] No orders found or invalid response');
      return [];
    } catch (e) {
      print('❌ [OrderService] Error fetching orders: $e');
      print('Stack trace: ${StackTrace.current}');
      return [];
    }
  }

  // Hủy đơn hàng
  static Future<bool> cancelOrder(String orderId) async {
    try {
      print('📦 [OrderService] Cancelling order: $orderId');
      
      final response = await ApiService.post(
        '${ApiConfig.orders}/$orderId/cancel',
        {},
        auth: true,
      );
      
      print('✅ [OrderService] Order cancelled');
      return response['success'] == true;
    } catch (e) {
      print('❌ [OrderService] Error cancelling order: $e');
      return false;
    }
  }
  
  // Cập nhật trạng thái đơn hàng (Admin)
  static Future<Order?> updateOrderStatus(String orderId, String status) async {
    try {
      print('📦 [OrderService] Updating order status: $orderId -> $status');
      
      final response = await ApiService.put(
        '${ApiConfig.orders}/$orderId/status?status=$status',
        {},
        auth: true,
      );
      
      if (response['success'] == true && response['data'] != null) {
        print('✅ [OrderService] Order status updated');
        return Order.fromJson(response['data']);
      }
      return null;
    } catch (e) {
      print('❌ [OrderService] Error updating order status: $e');
      return null;
    }
  }
}
